packages/platform-ios/src/commands/runIOS/index.ts

```diff
@@ -71,43 +71,51 @@ function runIOS(_: Array<string>, ctx: Config, args: FlagsT) {
     } "${chalk.bold(xcodeProject.name)}"`,
   );
 
-  const {device, udid} = args;
-
-  if (!device && !udid) {
+  // No need to load all available devices
+  if (!args.device && !args.udid) {
     return runOnSimulator(xcodeProject, scheme, args);
   }
 
+  if (args.device && args.udid) {
+    return logger.error(
+      'The `device` and `udid` options are mutually exclusive.',
+    );
+  }
+
   const devices = parseIOSDevicesList(
     child_process.execFileSync('xcrun', ['instruments', '-s'], {
       encoding: 'utf8',
     }),
   );
 
-  // first device is always the host Mac
-  if (devices.length <= 1) {
-    return logger.error('No iOS devices connected.');
-  }
-
-  const selectedDevice = matchingDevice(devices, device, udid);
-
-  if (selectedDevice) {
-    return runOnDevice(selectedDevice, scheme, xcodeProject, args);
-  }
-
-  if (device) {
-    return logger.error(
-      `Could not find a device named: "${chalk.bold(
-        String(device),
-      )}". ${printFoundDevices(devices)}`,
-    );
-  }
-
-  if (udid) {
-    return logger.error(
-      `Could not find a device with udid: "${chalk.bold(
-        udid,
-      )}". ${printFoundDevices(devices)}`,
-    );
+  if (args.udid) {
+    const device = devices.find(d => d.udid === args.udid);
+    if (!device) {
+      return logger.error(
+        `Could not find a device with udid: "${chalk.bold(
+          args.udid,
+        )}". ${printFoundDevices(devices)}`,
+      );
+    }
+    if (device.isSimulator) {
+      return runOnSimulator(xcodeProject, scheme, args);
+    } else {
+      return runOnDevice(device, scheme, xcodeProject, args);
+    }
+  } else {
+    const physicalDevices = devices.filter(d => !d.isSimulator);
+    if (physicalDevices.length === 0) {
+      return logger.error('No iOS devices connected.');
```

Comment:
What about host mac? Isn't it always displayed by `xcrun instruments -s`?

Reply:
Not after the incorrect change I made above :) I thought the above was unnecessarily returning the Mac only to remove it here.